src/librustc/ty/util.rs

```diff
@@ -295,10 +336,34 @@ impl<'tcx> TyCtxt<'tcx> {
     /// structure definitions.
     /// For `(Foo<Foo<T>>, Foo<dyn Trait>)`, the result will be `(Foo<T>, Trait)`,
     /// whereas struct_tail produces `T`, and `Trait`, respectively.
-    pub fn struct_lockstep_tails(self,
-                                 source: Ty<'tcx>,
-                                 target: Ty<'tcx>)
-                                 -> (Ty<'tcx>, Ty<'tcx>) {
+    ///
+    /// Must only be called after type-checking is complete; otherwise
+    /// normalization attempt may cause compiler bugs.
```

Comment:
Pretty sure you'd just get a cycle error? Or, rather, it's a bad idea to use this from anywhere that cares about lifetimes.

Reply:
Well I figured the fact that I witnessed an ICE when i called the wrong thing from rvalue_hint means that I can’t go wrong making the warning here super strong, right?

Reply:
What kind of ICE? Maybe it should just say "don't pass types with inference variables in, and don't use this if you want to keep your lifetimes".

Reply:
The distinction is important because e.g. the NLL borrowck uses inference despite being after typeck, and *cares about lifetimes even more*.

So the old "after typeck" should be rephrased across the compiler - it was a poor shorthand even back then, and now it's not even true anymore.

Reply:
In this [gist][] is some RUSTC_LOG and backtrace for the ICE that I was referring to up above

[gist]: https://gist.github.com/85b11805d1fca46dad0b03d0dd9a3abc

But sure, I am happy to revise the wording more.

Reply:
Yeah that's inference variables. Okay, minimal changes:
* "during type checking" -> "during type inference"
* "after type checking" -> "outside of type inference"